MarlinBook

JBenda opened the pull request:
BEGINRELEASENOTES

- Added Base for **MarlinBook**
   - allow the creation of objects for later easy parallel use
- Added fundamental test for the Module 
- Added DataModel UML (`/docs/diagrams/`)

ENDRELEASENOTES

- [x] check coding convention
- [x] adding exception handleing
- [x] adding titlet Histograms
- [x] replace tabs with spaes

Commit:
Forked Repository

* Added first iteration of BookStore API
  see include/marlin/book.h for details
* Improved API
* Added construct of BookStore

Commit:
Forked Reporsetory

Commit:
Added first iteration of BookStore API

see include/marlin/book.h for details

Commit:
Inmporved API

Commit:
Added construct of BookStore

Commit:
HistHnd fnPtr for diff behavior for diff flags

Commit:
Book function is now full Template

use Hnd<BookT> class specialisation for Type Handel mapping and validation

Added type trait

Commit:
Book function is now full Tepmlate

use Hnd<BookT> class spezialisation for Type Hnd mapping and validation

Commit:
addded type trait

Commit:
2 iteration for BookStore

* Fill function can be called
* Added  merger demo for Fill with Test
* Finished API Demo for Book Store
    * with small test
    * Merge is not Thread save !!(later)!!
    * Histogram Merge and reset is not math correct but ROOT7 don't provide a merge function
    * (*iterator) instead of iterator-> because ROOT7 bug (...)

Commit:
2 iteration for BookStore

fill function can be called

Commit:
Added  merger demo for Fille with Test

Commit:
Finished API Demo for Book Store

* with small test
* Merge is not Thread sava !!(later)!!
* Histogram Merge and reset is not math correct but ROOT7 don't provide a merge function
* (*itr) instead of itr-> becasue ROOT7 bug (...)

Commit:
moved test and resolved warnings

Commit:
Added MemLayout Demo

Commit:
Added MemLayout Demo

Commit:
Added RHist Test

Commit:
new book store iteration

Commit:
New book store iteration

* Added 3. Iteration of book store for histograms
* Documented and reformatted the BookStore code
* Formatted MemLayout.h

JBenda committed:
aded 2. iteraton of book store for histograms

JBenda committed:
polsish bookStore.h

JBenda committed:
Documented and reforamted the BookStore code

JBenda committed:
Formatted MemLayout.h

JBenda committed:
added some doxygen cosmetic

JBenda committed:
moved modifier handling in ModifierManager

JBenda committed:
Moved modifier handling in Modifier Manager

* use int hist for test (no floating-point error)

JBenda committed:
use int hist for test (no floatingpoint error)

JBenda committed:
Moved layer creation in ModMgr

JBenda committed:
Moved layer creation in ModMgr

JBenda committed:
Iterator 2 Handle Prototype implemented

Added book demo

JBenda committed:
Itr2 Handle Prototype implimented

JBenda committed:
Added book demo

JBenda committed:
added src/book to build path

only active when MARLIN_BOOK is set.

JBenda committed:
moving implimentation in .cc

JBenda committed:
moving implimentation in .cc

JBenda committed:
changed test :)

JBenda committed:
added MultiCopy Entry

JBenda committed:
added Multi Copy Entry

removed '.tmp' file from git

JBenda committed:
Changed test names

removed tmp file from git

JBenda committed:
Added Conditions and selections

* moved more in cpp files, many template explicit declaration needed
* romved unneeded files

JBenda committed:
Added Conditions and selections

* moved more in cpp files, many template explicit declaration needed

JBenda committed:
romved unneeded files

JBenda committed:
updatet test

JBenda committed:
Refactor Histogram Naming, and Root bridge

* introduce new File for every ROOT related stuff, for easier changes in future.
* Added Obj reference to HandleBase
* Added Shared Histogram Entry
* Added Namespace to ROOTAdapter
* Added Regex for Conditions
* Added remove funciton in BookStore

JBenda committed:
Refactor Histogram Naming, and Root bridge

* introduce new File for every ROOT related stuff, for easier changes in future.

JBenda committed:
Added Obj reference to HandleBase

JBenda committed:
Added Shared Histogram Entry

Commit:
Added Namespace to ROOTAdapter

Commit:
Added Regex for Conditions

Commit:
Added remove funciton in BookStore

Commit:
Formatted and write doxygen for Header fiels

* Formatted code in src/book
* fixed miss formatting

Commit:
Formatted and write doxygen for Header fiels

Commit:
Formatted code in src/book

Commit:
fixed miss formatting

Commit:
fixed doxygen

Commit:
stop doxygen from interpreting code files

Commit:
Added prototype for BookHelper

Commit:
Added prototype for BookHelper

Implemented BookHelper for Histograms
* added in BookStore base class
* added in Hist Book Helper for histograms

Usage of helper:
```cpp
	store.book<RH1F>("path","name").single()(axis);
	store.book<RH2F>("path", "name").multiCopy(8)(axis, axis);
```

Commit:
Implemented BookHelper for Histograms

* added in BookStore base class
* added in Hist Book Helper for histograms

Usage of helper:
```cpp
	store.book<RH1F>("path","name").single()(axis);
	store.book<RH2F>("path", "name").multiCopy(8)(axis, axis);
```

Commit:
a test , an test

Commit:
Formatting and documented BookHelper.

Commit:
Added test for accessing Entry from book store

Commit:
Added test for accessing Entry from book store

Commit:
Added easier access to handles from selection.

Commit:
Added /site/ to .gitignore

mkdocs saves in /site the compiled markdown, so there is no need to push this,
it's redundant with /docs

Commit:
Added short usage description to mkdocs

* new file:   docs/diagrams/BookStore-Data.html
  new file:   docs/diagrams/BookStore-Data.svg
	Dependency diagram for MarlinBook classes
* new file:   docs/user-manual/booking-system.md
  short introduction for the book Store

Commit:
Moved files to prepare rebase

* added implicit includes in BookStore.cc
* fixed test-book.cc

Commit:
Moved files to prepare rebase

* added implicit includes in BookStore.cc
* fixed test-book.cc

Commit:
added CMAKE for booking system

Commit:
Changed BookHelper to Entry Data

new syntax
```cpp
store.book("path", "name", EntryData<RH1F>(axis).single());
```

Commit:
moved parts in source files

Commit:
Added book to Doxyfile

Commit:
Formatted code and add doxygen

JBenda commented on doc/Doxyfile:
```diff
@@ -468,8 +468,7 @@ INPUT                  = $(DOX_INPUT)
 
-FILE_PATTERNS          = *.h \
-                         *.cc
+FILE_PATTERNS          = *.h
```
in .cc files are no public documentation, but template spezialisations and Macros, which Doxygen can't handle

rete replied:
No add it again. The source code is appended after the class documentation normally. If not, it should be and we should fix it then

JBenda commented on doc/mkdoxygen/Doxyfile:
```diff
@@ -459,7 +459,7 @@ WARN_LOGFILE           =
 # with spaces.
 
-INPUT                  = overview.txt ../../source/include ../../source/src ../../source/tinyxml
+INPUT                  = overview.txt ../../source/include ../../source/src ../../source/tinyxml ../../source/book
```
added MarlinBook to Doxygen

JBenda commented on doc/mkdoxygen/Doxyfile:
```diff
@@ -468,8 +468,7 @@ INPUT                  = overview.txt ../../source/include ../../source/src ../.
 
-FILE_PATTERNS          = *.h \
-                         *.cc
+FILE_PATTERNS          = *.h 
```
same as above
> in .cc files are no public documentation, but template specializations and Macros, which Doxygen can't handle

rete replied:
Same as above too

JBenda commented on source/include/marlin/Exceptions.h:
```diff
@@ -137,5 +137,6 @@ namespace marlin {
   MARLIN_DEFINE_EXCEPTION( SkipEventException ) ;
   MARLIN_DEFINE_EXCEPTION( StopProcessingException ) ;
   MARLIN_DEFINE_EXCEPTION( ParseException ) ;
+  MARLIN_DEFINE_EXCEPTION( BookStoreException );
```
Not in use, because of no clear idea of solving double booking.

rete replied:
We can iterate on this later on and make another pull request. Keep it for the time being

JBenda commented on test/tests/test-book-store.cc:
```diff
@@ -0,0 +1,219 @@
+	{
+
+		// EntrySingle entry = store.book<RH1F, RAxisConfig>("path", "name", {"a", 3, 1.0, 2.0}) ;	
+		EntrySingle entry = store.book("path", "name", EntryData<RH1F>({"a", 3, 1.0, 2.0}).single());
```
the two ways of booking, prefere 2.

JBenda committed:
update example

JBenda committed:
replace tabs with spaces

JBenda committed:
Added Histogram titles

rete commented on mkdocs.yml:
```diff
@@ -14,6 +14,7 @@ nav:
       - XML parser: 'user-manual/xml-parser.md'
       - Processor API: 'user-manual/processor-api.md'
       - Multi-threading with MarlinMT: 'user-manual/marlin-mt.md'
+      - Use global Objects in Parallel: 'user-manual/booking-system.md'
```
`Use global Objects in Parallel` -> `Histograms booking`

rete commented on source/CMakeLists.txt:
```diff
@@ -34,8 +35,12 @@ ENDIF()
     LIST( REMOVE_ITEM library_sources ./src/plugins/GearGeometry.cc )
 ENDIF()
+IF( MARLIN_BOOK )
+	# LIST(…) TODO when in framework
```
Still TODO ?

JBenda replied:
not any longer, CMakeLists.txt in `source/book/` should fix it

rete commented on .gitignore:
```diff
@@ -10,4 +10,5 @@
 /include/
 /lib/
 /*~
-/source/include/marlin/MarlinConfig.h
```
Why did you remove it ?

JBenda replied:
not on purpose, ... sorry :sweat:

rete commented on CMakeLists.txt:
```diff
@@ -141,7 +141,7 @@ INSTALL( FILES "${PROJECT_SOURCE_DIR}/source/python/compareMarlinSteeringFiles.p
 # -----------------------------------------------------------------------------
 
-
+set(MARLIN_BOOK ${MARLIN_BOOK} CACHE BOOL "not again" FORCE)
```
"not again" ?

JBenda replied:
sorry, debug line ...

rete commented on source/book/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+#################################
+# cmake file for building MarlinBOOK component
+# @author Remi Ete, DESY
```
Put your name their too

rete commented on source/book/include/marlin/book/Entry.h:
```diff
@@ -0,0 +1,183 @@
+    template < typename T >
+    struct trait {
+
+      /// Function to merge two objects of type T.
```
Do you need this trait class ?

JBenda replied:
I need a mapping from type to merge function ..., but, an overloaded function should do it also

rete commented on source/book/include/marlin/book/Entry.h:
```diff
@@ -0,0 +1,183 @@
+          return std::static_pointer_cast< EntryMultiShared< T > >( _entry )
+            ->handle() ;
+        }
+        // TODO add log
```
TODO ?

JBenda replied:
also, question, 
Should we log it if the creation failed because of a type mismatch or invalid Flag?
because when it's failed it's an internal error

rete commented on source/book/include/marlin/book/EntryData.h:
```diff
@@ -0,0 +1,62 @@
+      /// virtual Entry path
+      std::string path{""},
+        /// Entry name
+        name{""} ;
```
Declare the type here on a new line

rete commented on source/book/include/marlin/book/EntryData.h:
```diff
@@ -0,0 +1,62 @@
+        /// Entry name
+        name{""} ;
+      /// number of memory instances
+      std::size_t amt{0} ;
```
`amt` ? `nInstances` maybe ?

rete commented on source/book/include/marlin/book/MemLayout.h:
```diff
@@ -0,0 +1,129 @@
+
+    protected:
+      /// implementation from at
+      virtual std::shared_ptr< void > imp_at( std::size_t idx ) const = 0 ;
```
`impAt` to cope with conventions

rete commented on source/book/include/marlin/book/MemLayout.h:
```diff
@@ -0,0 +1,129 @@
+      /// implementation from at
+      virtual std::shared_ptr< void > imp_at( std::size_t idx ) const = 0 ;
+      /// implementation from merged
+      virtual std::shared_ptr< void > imp_merged() = 0 ;
```
`impMerged` to cope with conventions

rete commented on source/book/include/marlin/book/ROOTAdapter.h:
```diff
@@ -0,0 +1,80 @@
+
+      /// Generic Histogram.
+      template < int D, typename T, template < int, class > class... STAT >
+      using RH = ROOT::Experimental::RHist< D, T, STAT... > ;
```
`using RHist = ...` to cope with other typedefs

rete commented on source/book/include/marlin/book/ROOTAdapter.h:
```diff
@@ -0,0 +1,80 @@
+       *	@note can set with the CMAKE  Variable \cod {NAME}
+       */
+      constexpr std::size_t HistogramFillerBufferSize
+#ifdef HISTOGRAM_FILLER_BUFFER_SIZE
```
```cpp
#ifndef MARLIN_HIST_FILLER_BUFFER_SIZE
#define MARLIN_HIST_FILLER_BUFFER_SIZE 1024
#endif
constexpr std::size_t HistogramFillerBufferSize = MARLIN_HIST_FILLER_BUFFER_SIZE ;
```

rete commented on source/book/include/marlin/book/ROOTAdapter.h:
```diff
@@ -0,0 +1,80 @@
+       */
+      template < class T >
+      using RHistConcurrentFillManager
+        = ROOT::Experimental::RHistConcurrentFillManager< T, 1024 > ;
```
Use `HistogramFillerBufferSize` instead and after.

rete commented on source/book/include/marlin/book/Selection.h:
```diff
@@ -0,0 +1,120 @@
+     *	Used to doing action on a range of entries.
+     */
+    class Selection {
+      friend BookStore ;
```
Can't we get rid of this friend declaration ? We can expose a bit more the API if needed

JBenda replied:
we can, the only drawback the that the iteration constructer than is public, which should be no problem, because the definition is in the .cc file

rete commented on source/book/include/marlin/book/Selection.h:
```diff
@@ -0,0 +1,120 @@
+
+    public:
+      /// type for iteration through a Selection.
+      using iterator = typename std::vector< Entry >::const_iterator ;
```
call it `const_iterator` if it is a `const_iterator`

rete commented on source/book/include/marlin/book/ROOTAdapter.h:
```diff
@@ -0,0 +1,80 @@
+#include "ROOT/RHist.hxx"
+#include "ROOT/RHistConcurrentFill.hxx"
+#include "ROOT/RHistData.hxx"
+#include "ROOT/span.hxx"
```
Include RSpan instead

rete commented on source/book/include/marlin/book/ROOTAdapter.h:
```diff
@@ -0,0 +1,80 @@
+
+      /// Histogram merge function.
+      template < typename TO, typename FROM >
+      void ADD( const TO &to, const FROM &from ) {
```
`ADD` -> `addHists`

rete commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -0,0 +1,155 @@
+#include <unordered_map>
+
+// -- map ROOT to MarlinBook types
+#include "marlin/book/ROOTAdapter.h"
```
Group the includes here

rete commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -0,0 +1,155 @@
+    template <>
+    class EntryDataBase< void > {
+    public:
+      EntryDataBase()                        						= default ;
```
Fix the indent here

rete commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -0,0 +1,155 @@
+      auto book( const std::string_view &path,
+                 const std::string_view &name,
+                 const T &               data ) {
+        // TODO static assert and type check
```
TODO ?

JBenda replied:
Not any longer, because of the of the book function wich is called. Should be save.

rete commented on source/book/include/marlin/book/Condition.h:
```diff
@@ -0,0 +1,152 @@
+       */
+      bool operator()( const EntryKey &key ) const { return _fiterFn( key ); }
+
+      // TODO: 'and' and 'or' are keywords in c++(and = &&, or = ||)
```
TODO ?

JBenda replied:
more an discuss, is Not, And, Or ok for the functions? because it's not the naming convention. When we overload the `&&` `||` operator `condition.and(condition2)` is valid ...

rete commented on source/book/include/marlin/book/Entry.h:
```diff
@@ -0,0 +1,183 @@
+       *	check if there is a reference stored or not.
+       * 	@return true if entry is valid.
+       */
+      bool valid() const { return static_cast< bool >( _entry ); }
```
Use nullptr comparison to make it explicit

rete commented on source/book/include/marlin/book/EntryData.h:
```diff
@@ -0,0 +1,62 @@
+      /// default constructor.
+      EntryKey() : type{std::type_index( typeid( void ) )} {}
+
+      // TODO: clean solution
```
TODO ?

JBenda replied:
not any longer, sorry, but may ugly

rete commented on source/book/include/marlin/book/Handle.h:
```diff
@@ -0,0 +1,60 @@
+     */
+    template < typename T >
+    class BaseHandle {
+      std::shared_ptr< MemLayout > _mem ;
```
Members documentation

rete commented on source/book/include/marlin/book/Hist.h:
```diff
@@ -0,0 +1,465 @@
+       *	@note \code {RHist::GetEntries()} is broken after that.
+       */
+      static void
+      Merge( const std::shared_ptr< types::RH< D, T, STAT... > > &dst,
```
`merge` to stick to conventions

rete commented on source/book/src/BookStore.cc:
```diff
@@ -0,0 +1,187 @@
+
+    //--------------------------------------------------------------------------
+
+    template < class T, typename... Args_t >
```
All templates in header files

JBenda replied:
this is only the definition, the declaration is in the header file [`BookStore.h`](https://github.com/JBenda/MarlinMT/blob/8c5e0a393c7b0c9532e3c98a87ec5eadb1ccf03e/source/book/include/marlin/book/BookStore.h#L94)

rete commented on source/book/src/Hist.cc:
```diff
@@ -0,0 +1,274 @@
+  namespace book {
+    using namespace types ;
+
+    template < int D, typename T, template < int, class > class... STAT >
```
This should also go in the header file

JBenda replied:
because constructor should be every time in the header ?, if yes i must change a few more things.

rete commented on source/book/src/Selection.cc:
```diff
@@ -0,0 +1,88 @@
+
+    Selection::Selection( const Selection &sel,
+                          const Condition &cond,
+                          ComposeStrategie strategie )
```
Strategie -> Strategy

rete commented on test/CMakeLists.txt:
```diff
@@ -17,6 +17,12 @@ if( MARLIN_LCIO )
+if( MARLIN_BOOK )
+	#find_package(MarlinBook CONFIG REQUIRED )
+	include_directories( ${PROJECT_SOURCE_DIR}/source/book/include)
+	link_libraries( "${PROJECT_SOURCE_DIR}/build/lib/libMarlinBook.so" )
```
```cmake
link_libraries( MarlinBook )
```

rete commented on test/tests/test-book-store.cc:
```diff
@@ -0,0 +1,223 @@
+#include <string>
+
+template<typename T>
+void MergeaHist(const std::shared_ptr<T>& dst, const std::shared_ptr<T>& src) {
```
Use your function implemented in the framework

rete commented on test/tests/test-book-store.cc:
```diff
@@ -0,0 +1,223 @@
+
+std::string mergedUnicStr() {
+  static std::size_t num = 0;
+  return std::to_string(++num);
```
WARNING: Fine here, but std::to_string is not thread-safe

JBenda replied:
ok, thanks for the note

rete commented on test/tests/test-mem-layout.cc:
```diff
@@ -0,0 +1,75 @@
+	}
+}
+template<typename ... Args_t>
+class SharedType1 : public SharedMemLayout<Type1, MergeType1, Args_t ... > {
```
Why to you need to sub-class here ?

JBenda replied:
for binding the merge function with the class (for Type1), we don't really need it. But the notation, this way is more likely used when you used the MemLayout, because the alternative is to construct it with:
`SharedMemLayout<Type1, MergeType1, int, int, int, int, int> sMem(…)`

rete commented on test/tests/test-mem-layout.cc:
```diff
@@ -0,0 +1,75 @@
+	SharedType1 (std::size_t num_inst, Args_t ... args) : SharedMemLayout<Type1, MergeType1, Args_t ... >{num_inst, args ...}{}
+};
+
+void MergeHist(const std::shared_ptr<RH1I>& p1, const std::shared_ptr<RH1I>& p2) {
```
Use your nice framework here

rete commented:
General comments:
- Check your indent. It is still not correct in a few places
- The expression AMT doesn't look good. We need to sit down and find a new expression for it
- Check the naming conventions, in particular for upper/lower case function name

Else it looks good ! Nice job !

JBenda committed:
Rename tutorial and re add *.cpp files for doxygen

JBenda committed:
refactor make files

JBenda committed:
removed old help functions

JBenda committed:
Fixes for pull-request
BookStore.h:
+ regroup includes
+ removed old TODO
+ fixed indents

Entry.h:
+ use `shared_ptr == nullptr`, instead of `bool cast`

EntryData.h:
+ removed old TODO
+ reformat member declaration

Handle.h:
+ add member documentation for `BaseHandle`

MemLayout.h:
+ rename internal functions

ROOTAdapter.hpp:
+ use `ROOT::RSpan` instead of `ROOT::span` for fix dependencies

Selection.h/cpp:
+ removed friend to BookStore
+ fixed spelling

JBenda committed:
finished renames

JBenda committed:
added exceptions

JBenda committed:
changed erros to Marlin intern errors

JBenda committed:
Moved Template function definition in header

* pro: no implicit declaration
* contra: more compile time

JBenda committed:
explicit AddFunction call

JBenda committed:
Minor changes

* changed trait struct with function (for merging objects)
* changed Condition functions to operator (&, |, !)

JBenda committed:
removed tabs

JBenda committed:
redo changes

JBenda committed:
added merge function definition

JBenda committed:
removed semi-collum

rete commented on docs/user-manual/booking-system.md:
```diff
@@ -0,0 +1,144 @@
+The `BookStore` is an Object which managed Histograms for later universal access.
```
Can you move this file to a new directory `internal-documentation` instead of `user-manual` ?

JBenda committed:
Moved mkdocs `booking-system` entry

from `user-manual` to `internal-documentation`

JBenda committed:
Cleaning

* fixed spelling in `docs/internal-documentation/booking-system.md`
* removed deprecated files
   + book/src/BookStore.cc~HEAD
   + include/marlin/book/Modifier.h

JBenda committed:
Removed empty line diffs to master

JBenda committed:
revert indent replacment
